src/os/bluestore/BlueFS.cc

```diff
@@ -3662,21 +3663,58 @@ int BlueFS::truncate(FileWriter *h, uint64_t offset)/*_WF_L*/
     if (r < 0)
       return r;
   }
-  if (offset == h->file->fnode.size) {
-    return 0;  // no-op!
-  }
-  if (offset > h->file->fnode.size) {
+  if (offset > fnode.size) {
     ceph_abort_msg("truncate up not supported");
   }
-  ceph_assert(h->file->fnode.size >= offset);
+  ceph_assert(offset <= fnode.size);
   _flush_bdev(h);
-
-  std::lock_guard ll(log.lock);
-  vselector->sub_usage(h->file->vselector_hint, h->file->fnode.size);
-  h->file->fnode.size = offset;
-  h->file->is_dirty = true;
-  vselector->add_usage(h->file->vselector_hint, h->file->fnode.size);
-  log.t.op_file_update_inc(h->file->fnode);
+  {
+    std::lock_guard ll(log.lock);
+    std::lock_guard dl(dirty.lock);
+    bool changed_extents = false;
+    vselector->sub_usage(h->file->vselector_hint, fnode);
+    uint64_t x_off = 0;
+    auto p = fnode.seek(offset, &x_off);
+    uint64_t cut_off =
+      (p == fnode.extents.end()) ? 0 : p2roundup(x_off, alloc_size[p->bdev]);
+    uint64_t new_allocated;
+    if (0 == cut_off) {
+      // whole pextent to remove
+      changed_extents = true;
```

Comment:
No, because cutting tail for an extent is also a change.

Reply:
Not sure I'm getting this point.
When p ==fnode.extents.end() (and hence cut_off  == 0) there is nothing to cut at the tail IMO. And hence there is no need to update the log etc... I can only imagine one case - when truncate effectively expands the fnode up to allocated boundary. And one needs to update fnode.size then. Do you mean this case?

Reply:
I might be missing something.
```
AAAABBBBCCCCDDDDEEEE
    ^case 1, remove B-E
AAAABBBBCCCCDDDDEEEE
     ^case 2, cut B, remove C-E
AAAABBBBCCCCDDDDEEEE
       ^case 3, whole B remains, remove C-E         
```

So in case 1,  cut_off == 0 but p!=fnode.extents.end().

The problem is different, and I think I get the confusion.
In case of `p == fnode.extents.end()` the whole extent dropping part should be skipped altogether and no log updated.
@ifed01 ?

Reply:
Right, that's what I meant